File: examples/part7-vars-go/vars.go
// Package main — vars: the expression interpreter grown into a tiny language
// with variables and a persistent environment.
//
// Lesson 1 evaluated one expression. Real languages remember things: you name a
// value and use it later. That is a variable, and the thing holding the names is
// an environment — a name->value map threaded through evaluation. Add that and
// the calculator becomes a language. Still integer arithmetic, so results stay
// byte-identical across the three languages. Library and command share one
// package so `go run .` and `go test` both work.
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

// Env is the running memory of the program: variable names to their values.
type Env map[string]int64

// ---------------------------------------------------------------- lexer

// Token: Kind 'n' number (Num), 'i' identifier (Name), '=' assign, else the
// operator/parenthesis character.
type Token struct {
	Kind byte
	Num  int64
	Name string
}

// Tokenize turns one statement's source into tokens, adding identifiers and the
// '=' assignment token to the lexer from lesson 1.
func Tokenize(src string) ([]Token, error) {
	r := []rune(src)
	var tokens []Token
	i := 0
	for i < len(r) {
		c := r[i]
		switch {
		case c == ' ' || c == '\t' || c == '\r' || c == '\n':
			i++
		case c == '=' || c == '+' || c == '-' || c == '*' || c == '/' || c == '(' || c == ')':
			tokens = append(tokens, Token{Kind: byte(c)})
			i++
		case c >= '0' && c <= '9':
			start := i
			for i < len(r) && r[i] >= '0' && r[i] <= '9' {
				i++
			}
			n, err := strconv.ParseInt(string(r[start:i]), 10, 64)
			if err != nil {
				return nil, fmt.Errorf("number too large: %s", string(r[start:i]))
			}
			tokens = append(tokens, Token{Kind: 'n', Num: n})
		case unicode.IsLetter(c) || c == '_':
			start := i
			for i < len(r) && (unicode.IsLetter(r[i]) || unicode.IsDigit(r[i]) || r[i] == '_') {
				i++
			}
			tokens = append(tokens, Token{Kind: 'i', Name: string(r[start:i])})
		default:
			return nil, fmt.Errorf("unexpected character '%c'", c)
		}
	}
	return tokens, nil
}

// ---------------------------------------------------------------- parser

// Expr: Kind 'n' number (Num), 'v' variable (Name), 'u' unary neg (L), else a
// binary operator (L, R).
type Expr struct {
	Kind byte
	Num  int64
	Name string
	L, R *Expr
}

// Stmt: Kind 'a' assignment (Name = Expr) or 'e' bare expression.
type Stmt struct {
	Kind byte
	Name string
	Expr *Expr
}

type parser struct {
	tokens []Token
	pos    int
}

func (p *parser) peek() (Token, bool) {
	if p.pos < len(p.tokens) {
		return p.tokens[p.pos], true
	}
	return Token{}, false
}

func (p *parser) next() (Token, bool) {
	t, ok := p.peek()
	if ok {
		p.pos++
	}
	return t, ok
}

func (p *parser) expr() (*Expr, error) {
	left, err := p.term()
	if err != nil {
		return nil, err
	}
	for {
		t, ok := p.peek()
		if !ok || (t.Kind != '+' && t.Kind != '-') {
			return left, nil
		}
		p.pos++
		right, err := p.term()
		if err != nil {
			return nil, err
		}
		left = &Expr{Kind: t.Kind, L: left, R: right}
	}
}

func (p *parser) term() (*Expr, error) {
	left, err := p.factor()
	if err != nil {
		return nil, err
	}
	for {
		t, ok := p.peek()
		if !ok || (t.Kind != '*' && t.Kind != '/') {
			return left, nil
		}
		p.pos++
		right, err := p.factor()
		if err != nil {
			return nil, err
		}
		left = &Expr{Kind: t.Kind, L: left, R: right}
	}
}

// factor := Num | Ident | '(' expr ')' | '-' factor
func (p *parser) factor() (*Expr, error) {
	t, ok := p.next()
	if !ok {
		return nil, fmt.Errorf("unexpected end of input")
	}
	switch t.Kind {
	case 'n':
		return &Expr{Kind: 'n', Num: t.Num}, nil
	case 'i':
		return &Expr{Kind: 'v', Name: t.Name}, nil
	case '-':
		inner, err := p.factor()
		if err != nil {
			return nil, err
		}
		return &Expr{Kind: 'u', L: inner}, nil
	case '(':
		inner, err := p.expr()
		if err != nil {
			return nil, err
		}
		if rp, ok := p.next(); !ok || rp.Kind != ')' {
			return nil, fmt.Errorf("expected ')'")
		}
		return inner, nil
	default:
		return nil, fmt.Errorf("unexpected token '%c'", t.Kind)
	}
}

// ParseStmt parses one statement: `name = expr` is an assignment (identifier
// followed by '='); anything else is a bare expression.
func ParseStmt(tokens []Token) (Stmt, error) {
	isAssign := len(tokens) >= 2 && tokens[0].Kind == 'i' && tokens[1].Kind == '='
	p := &parser{tokens: tokens}
	var stmt Stmt
	if isAssign {
		name := tokens[0].Name
		p.pos = 2 // skip identifier and '='
		e, err := p.expr()
		if err != nil {
			return Stmt{}, err
		}
		stmt = Stmt{Kind: 'a', Name: name, Expr: e}
	} else {
		e, err := p.expr()
		if err != nil {
			return Stmt{}, err
		}
		stmt = Stmt{Kind: 'e', Expr: e}
	}
	if p.pos != len(p.tokens) {
		return Stmt{}, fmt.Errorf("unexpected trailing input: '%c'", p.tokens[p.pos].Kind)
	}
	return stmt, nil
}

// ---------------------------------------------------------------- evaluator

// Eval evaluates an expression against the environment; an unbound variable is
// an error.
func Eval(e *Expr, env Env) (int64, error) {
	switch e.Kind {
	case 'n':
		return e.Num, nil
	case 'v':
		v, ok := env[e.Name]
		if !ok {
			return 0, fmt.Errorf("undefined variable '%s'", e.Name)
		}
		return v, nil
	case 'u':
		v, err := Eval(e.L, env)
		return -v, err
	default:
		a, err := Eval(e.L, env)
		if err != nil {
			return 0, err
		}
		b, err := Eval(e.R, env)
		if err != nil {
			return 0, err
		}
		switch e.Kind {
		case '+':
			return a + b, nil
		case '-':
			return a - b, nil
		case '*':
			return a * b, nil
		case '/':
			if b == 0 {
				return 0, fmt.Errorf("division by zero")
			}
			return a / b, nil
		}
		return 0, fmt.Errorf("unreachable")
	}
}

// Exec runs one statement, returning its value. An assignment stores the value
// and evaluates to it; a bare expression just evaluates.
func Exec(stmt Stmt, env Env) (int64, error) {
	if stmt.Kind == 'a' {
		v, err := Eval(stmt.Expr, env)
		if err != nil {
			return 0, err
		}
		env[stmt.Name] = v
		return v, nil
	}
	return Eval(stmt.Expr, env)
}

// RunProgram runs a whole program: one statement per non-empty line, sharing a
// single environment so state persists. Returns one "line  =>  value" (or error)
// string per statement.
func RunProgram(src string) []string {
	env := Env{}
	var out []string
	for _, line := range strings.Split(src, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" {
			continue
		}
		v, err := runOne(trimmed, env)
		if err != nil {
			out = append(out, fmt.Sprintf("%s  =>  error: %s", trimmed, err))
		} else {
			out = append(out, fmt.Sprintf("%s  =>  %d", trimmed, v))
		}
	}
	return out
}

func runOne(src string, env Env) (int64, error) {
	tokens, err := Tokenize(src)
	if err != nil {
		return 0, err
	}
	stmt, err := ParseStmt(tokens)
	if err != nil {
		return 0, err
	}
	return Exec(stmt, env)
}

File: examples/part7-vars-py/vars.py
"""vars — the expression interpreter grown into a tiny language with memory.

Lesson 1 evaluated one expression. Real languages remember things: you name a
value and use it later. That is a variable, and the thing holding the names is an
*environment* — a name->value map threaded through evaluation. Add that and the
calculator becomes a language. Still integer arithmetic, so results stay exact
and byte-identical across the three languages. The environment introduced here is
the seed of scopes and closures in the lessons to come.
"""

from __future__ import annotations

from dataclasses import dataclass

Env = dict[str, int]


class CalcError(Exception):
    """A lexing, parsing, or evaluation error. Its message is user-facing."""


# ---------------------------------------------------------------- lexer


@dataclass(frozen=True)
class Token:
    """Kind ``'n'`` number (in ``num``), ``'i'`` identifier (in ``name``),
    ``'='`` assign, else the operator/parenthesis character."""

    kind: str
    num: int = 0
    name: str = ""


def tokenize(src: str) -> list[Token]:
    """Turn one statement's source into tokens, adding identifiers and the ``=``
    assignment token to the lexer from lesson 1."""
    tokens: list[Token] = []
    i = 0
    while i < len(src):
        c = src[i]
        if c in " \t\r\n":
            i += 1
        elif c in "=+-*/()":
            tokens.append(Token(c))
            i += 1
        elif c.isdigit():
            start = i
            while i < len(src) and src[i].isdigit():
                i += 1
            tokens.append(Token("n", num=int(src[start:i])))
        elif c.isalpha() or c == "_":
            start = i
            while i < len(src) and (src[i].isalnum() or src[i] == "_"):
                i += 1
            tokens.append(Token("i", name=src[start:i]))
        else:
            raise CalcError(f"unexpected character '{c}'")
    return tokens


# ---------------------------------------------------------------- parser


@dataclass(frozen=True)
class Expr:
    """Kind ``'n'`` number (``num``), ``'v'`` variable (``name``), ``'u'`` unary
    neg (``left``), else a binary operator (``left``, ``right``)."""

    kind: str
    num: int = 0
    name: str = ""
    left: Expr | None = None
    right: Expr | None = None


@dataclass(frozen=True)
class Stmt:
    """Kind ``'a'`` assignment (``name = expr``) or ``'e'`` bare expression."""

    kind: str
    expr: Expr
    name: str = ""


class _Parser:
    def __init__(self, tokens: list[Token]) -> None:
        self.tokens = tokens
        self.pos = 0

    def _peek(self) -> Token | None:
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def _next(self) -> Token | None:
        t = self._peek()
        if t is not None:
            self.pos += 1
        return t

    def expr(self) -> Expr:
        left = self.term()
        while (t := self._peek()) is not None and t.kind in "+-":
            self.pos += 1
            left = Expr(t.kind, left=left, right=self.term())
        return left

    def term(self) -> Expr:
        left = self.factor()
        while (t := self._peek()) is not None and t.kind in "*/":
            self.pos += 1
            left = Expr(t.kind, left=left, right=self.factor())
        return left

    def factor(self) -> Expr:  # factor := Num | Ident | '(' expr ')' | '-' factor
        t = self._next()
        if t is None:
            raise CalcError("unexpected end of input")
        if t.kind == "n":
            return Expr("n", num=t.num)
        if t.kind == "i":
            return Expr("v", name=t.name)
        if t.kind == "-":
            return Expr("u", left=self.factor())
        if t.kind == "(":
            inner = self.expr()
            close = self._next()
            if close is None or close.kind != ")":
                raise CalcError("expected ')'")
            return inner
        raise CalcError(f"unexpected token '{t.kind}'")


def parse_stmt(tokens: list[Token]) -> Stmt:
    """Parse one statement: ``name = expr`` is an assignment (identifier followed
    by ``=``); anything else is a bare expression."""
    is_assign = len(tokens) >= 2 and tokens[0].kind == "i" and tokens[1].kind == "="
    p = _Parser(tokens)
    if is_assign:
        name = tokens[0].name
        p.pos = 2  # skip identifier and '='
        stmt = Stmt("a", expr=p.expr(), name=name)
    else:
        stmt = Stmt("e", expr=p.expr())
    if p.pos != len(p.tokens):
        raise CalcError(f"unexpected trailing input: '{p.tokens[p.pos].kind}'")
    return stmt


# ---------------------------------------------------------------- evaluator


def evaluate(e: Expr, env: Env) -> int:
    """Evaluate an expression against the environment; an unbound variable is an
    error."""
    if e.kind == "n":
        return e.num
    if e.kind == "v":
        if e.name not in env:
            raise CalcError(f"undefined variable '{e.name}'")
        return env[e.name]
    if e.kind == "u":
        assert e.left is not None
        return -evaluate(e.left, env)
    assert e.left is not None and e.right is not None
    a, b = evaluate(e.left, env), evaluate(e.right, env)
    if e.kind == "+":
        return a + b
    if e.kind == "-":
        return a - b
    if e.kind == "*":
        return a * b
    if e.kind == "/":
        if b == 0:
            raise CalcError("division by zero")
        q = abs(a) // abs(b)  # truncate toward zero, not Python's floor //
        return -q if (a < 0) != (b < 0) else q
    raise CalcError("unreachable")  # pragma: no cover


def execute(stmt: Stmt, env: Env) -> int:
    """Run one statement, returning its value. An assignment stores the value and
    evaluates to it; a bare expression just evaluates."""
    if stmt.kind == "a":
        v = evaluate(stmt.expr, env)
        env[stmt.name] = v
        return v
    return evaluate(stmt.expr, env)


def run_program(src: str) -> list[str]:
    """Run a whole program: one statement per non-empty line, sharing a single
    environment so state persists. Returns one ``"line  =>  value"`` (or error)
    string per statement."""
    env: Env = {}
    out: list[str] = []
    for line in src.split("\n"):
        trimmed = line.strip()
        if not trimmed:
            continue
        try:
            v = execute(parse_stmt(tokenize(trimmed)), env)
            out.append(f"{trimmed}  =>  {v}")
        except CalcError as e:
            out.append(f"{trimmed}  =>  error: {e}")
    return out


if __name__ == "__main__":
    import sys

    arg = " ".join(sys.argv[1:]).strip()
    src = arg.replace(";", "\n") if arg else sys.stdin.read()
    for line in run_program(src):
        print(line)
